sdk/ai/ai-inference-rest/src/tracingPolicy.ts

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import {
+  type TracingClient,
+  type TracingContext,
+  type TracingSpan,
+  createTracingClient,
+} from "@azure/core-tracing";
+import { logger } from "./logger.js";
+import { getErrorMessage } from "@azure/core-util";
+import {
+  tryProcessResponse,
+  onStartTracing,
+  tryProcessError,
+  getSpanName,
+  getRequestBody,
+} from "./tracingHelper.js";
+import {
+  PipelinePolicy,
+  PipelineRequest,
+  PipelineResponse,
+  SendRequest,
+} from "@azure/core-rest-pipeline";
+import { SDK_VERSION } from "./constants.js";
+
+/**
+ * The programmatic identifier of the tracingPolicy.
+ */
+export const tracingPolicyName = "inferenceTracingPolicy";
+
+/**
+ * A simple policy to create OpenTelemetry Spans for each request made by the pipeline
+ * that has SpanOptions with a parent.
+ * Requests made without a parent Span will not be recorded.
+ */
+export function tracingPolicy(): PipelinePolicy {
+  const tracingClient = createTracingClient({
+    namespace: "Microsoft.CognitiveServices",
+    packageName: "@azure/ai-inference-rest",
+    packageVersion: SDK_VERSION,
+  });
+
+  return {
+    name: tracingPolicyName,
+    async sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse> {
+      const url = new URL(request.url);
+      if (
+        !tracingClient ||
+        !url.href.endsWith("/chat/completions") ||
+        getRequestBody(request)?.body?.stream
+      ) {
+        return next(request);
+      }
+
+      const { span, tracingContext } = tryCreateSpan(tracingClient, request) ?? {};
+
+      if (!span || !tracingContext) {
+        return next(request);
+      }
+
+      try {
+        request.tracingOptions ??= {};
+        request.tracingOptions.tracingContext = tracingContext;
+
+        onStartTracing(span, request, request.url);
+        const response = await tracingClient.withContext(tracingContext, next, request);
+        tryProcessResponse(span, response);
+        return response;
+      } catch (err: any) {
+        tryProcessError(span, err);
+        throw err;
+      } finally {
+        span.end();
+      }
+    },
+  };
+}
+
+function tryCreateSpan(
+  tracingClient: TracingClient,
+  request: PipelineRequest,
+): { span: TracingSpan; tracingContext: TracingContext } | undefined {
+  try {
+    // As per spec, we do not need to differentiate between HTTP and HTTPS in span name.
+    const { span, updatedOptions } = tracingClient.startSpan(
+      getSpanName(request),
+      { tracingOptions: request.tracingOptions },
+      {
+        spanKind: "client",
+      },
+    );
+
+    // If the span is not recording, don't do any more work.
+    if (!span.isRecording()) {
+      span.end();
```

Comment:
This actually is the pattern for tracing in this repository.   When users fail or did not register instrumentation, isRecording will be false, and we don't trace.

Reply:
I hope we'll fix it at some point - https://github.com/Azure/azure-sdk-for-js/issues/31195

Could you please update this PR so that it already does the right thing when the rest of the repo is updated?

Thanks!

Reply:
But tracingClient in core-tracing doesn't have isEnabled.   When should we enable or disable is another question.      @maorleger  can you share your input?

Reply:
just don't call isEnabled

Reply:
@lmolkova    Sorry..   I need more clarification.   If I don't check isEnabled, I will have to call tracingClient.startSpan, tracingClient.withContext, span.end() always.    Currently my code always call tracingClient.startSpan and span.end() always.   So why are you asking me to call  tracingClient.withContext?    I don't need to call withContext to propergate to core-rest-client because my code call next(request) that will move on to the core-rest-client to instrument HTTP span.

Reply:
the only think I'm asking for is to change behavior around `isRecording`.

You don't need to add any attributes or events if span is not recording, but it should keep going till the end of operation in the same way as recording span does. 

I.e. you'd write code like

```cs
span = tracer.startSpan(name, kind, startAttributes)
if (span.isRecording()) {
  // add more attributes and events
}

try {
  // call original method
} catch (Exception ex) {
  span.setStatus(ex, ex. Message);
  if (span.isRecording()) {
    span.setAttribute("error.type", ex.FullName);
  }
} finally {
  span.end();
}
```

Reply:
I just did what you said.    Here is what happens:
1. if there is exception in the catch will set attribute for error type only and status. 
2. if there is isRecording is false, it will call tracingClient.withContext that will set tracing context to span.

I still don't understand why we should set status, error type attribute, and tracing context for a span that doesn't have instrumentation registration.
@maorleger   could you share your thought?

Reply:
@lmolkova is the expert here, so I will defer to her regarding guidance. To be fair to you @howieleung our tracingPolicy _does_ do the same thing (end and return early) if the span is not recording and that has been there for 3 years or so. With that said, we'll get that fixed up in core if that's what Liudmila recommends.

As far as curiosity getting the better of me I took a look at https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-sdk-trace-base/src/Span.ts to better understand the implementation (this is in sdk-trace-base, but sdk-trace-node uses this span implementation as well)

`isRecording` and `isSpanEnded` both use `this._ended` boolean

If the span is ended, calling `end` on it is essentially a `no-op`: https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-sdk-trace-base/src/Span.ts#L252-L258 

The rest of the methods just return early if the span is ended so they're safe to call https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-sdk-trace-base/src/Span.ts#L228

It's _weird_ to me that `isRecording` is based off the value of `this._ended` - a non-recording span hasn't technically ended until `end` is called as far as I understand. If a non-recording span is the same as an ended span then it doesn't necessarily matter when we call `end` on it.

For why it's like this in core - my guess is back then OTel was not very stable and we were concerned about tracing issues impacting the entire operation so we wanted to be _super_ defensive in tracingPolicy and preferred dropping some traces to failing the request altogether. I bet we can remove a lot of that boilerplate now that OTel is largely stable.

Reply:
Pragmatically though, would it be acceptable to merge your PR as is and create a follow-up issue for you to investigate? This doesn't feel blocking to me but I am happy to defer to y'all on that

Reply:
@maorleger   thanks for checking the otel code these days.   After all I feel that we should tracingSpan should not expose isRecording function.   Our sdk should not check isRecording.  We should just call setAttribute, addEvent, and setStatus.   Then let otel to check this._ended.   If it is ended, just return.   I am gonna create an issue on this.

Reply:
Opened issue https://github.com/Azure/azure-sdk-for-js/issues/31206